Improved http responses

soren5 opened the pull request:
Implemented error codes 400 401 429 and 404 in Handler.php. Also separated Exception handling into API and Web to avoid previously confusing responses.
Tested utilizing this Postman collection:
 https://www.getpostman.com/collections/ce1c3a433911f25a6063

soren5 committed:
chore: add .vscode to .gitignore

Add visual code configuration files to .gitignore.

soren5 committed:
feat: Add error code 400

- Separate exception handling into api and non api responses.
- Handle NotFoundHttpException generated by invalid URL syntax
with appropriate error message.

soren5 committed:
feat: add error code 401

-Handle AuthenticationException thrown by invalid acces token
and return appropriate response.

soren5 committed:
feat:  add error code 429

-Handle exception thrown by exceeding rate limit
- Create fail safe for unexpected exception

soren5 committed:
feat: add error code 404

- Reintroduce response for error code 404 with more detailed logging

soren5 committed:
fix: add missing dependencies

add missing dependencies for AuthenticationException and HttpException

soren5 committed:
chore: associate with issues

- fix #7
- fix #5

hmatalonga commented on app/Exceptions/Handler.php:
```diff
@@ -49,18 +51,43 @@ public function report(Exception $exception)
     {
-        if (! $request->expectsJson()) {
-            return parent::render($request, $exception);
+        if (strpos($request->getUri(), '/api/v') !== false) {
```
Instead of `strpos` use `$request->is()` since it is a helper method from the framework itself and allows us to use simple regex.

https://laravel.com/docs/5.5/requests#request-path-and-method

Also I think would be good to extract this conditional to a helper method, later we may need to add more constraints for what is an api method... e.g. `isApiMethod($request)`.

hmatalonga commented on app/Exceptions/Handler.php:
```diff
@@ -49,18 +51,43 @@ public function report(Exception $exception)
-                'message' => 'Resource not found.',
-            ], 404);
+        else{
+            return parent::render($request, $exception);
```
I would recommend to invert the order of the statements, this will clean up the code a little bit, something along the lines of the original code:
```php
if (! $request->isApiMethod()) {
   return parent::render($request, $exception);
}

if ... // All the rest
```

The idea is to check the negation first to avoid extra nested conditionals.

This is a neat refactoring tip I learned from the Laravel community :+1:

hmatalonga commented on app/Exceptions/Handler.php:
```diff
@@ -49,18 +51,43 @@ public function report(Exception $exception)
+                    'message' => 'Unable to authenticate request, check if your access token is correct.',
+                ], 401);
+            } elseif ($exception instanceof HttpException) {
+                if ($exception->getMessage() == 'Too Many Attempts.') {
```
Switch to using the `$exception` code (if possible) instead of the message, since is less likely to change over time. :wink:

hmatalonga replied:
So possibly the exception code is not always unique for each class... 

However we still can avoid using the messages, since there are exceptions for each concrete case, check them here:
https://api.symfony.com/3.4/Symfony/Component/HttpKernel/Exception.html

I suppose by adding more switch statements we will solve the problem.

A few of these exceptions extend from some more generic ones like ` TooManyRequestsHttpException` extends from `HttpException` that's why the error is caught there too.

soren5 replied:
I've been having a hard time reproducing the TooManyRequestsHttpException, I eventually found this:
https://github.com/laravel/framework/issues/13623
It's mentioned that Laravel does not use TooManyRequestsHttpException but I can't find anything else that confirms this other than my experience trying to catch it.
How should I proceed? Is there another alternative other than using getMessage()?
EDIT: Found an alternative further down the thread I linked, will be using getStatusCode()

hmatalonga replied:
I wasn't aware that they don't use the `TooManyRequestsHttpException`.

So as mentioned in the Laravel thread, the two possible solutions would be to create a custom middleware to handle the ThrottleRequests or to use the `$exception->getStatusCode() == 429` as you said.

To keep everything in the same scope I think it would suffice to use the status code approach since we don't need to override too much stuff.

hmatalonga commented on app/Exceptions/Handler.php:
```diff
@@ -49,18 +51,43 @@ public function report(Exception $exception)
+            } elseif ($exception instanceof ModelNotFoundException) {
+                //This should be revised if routes are changed
+                $pathArray = explode("/", $request->getUri());
+                $objectId = explode("?", $pathArray[6])[0];
```
We should improve this approach I believe. Consider this test case, an endpoint `api/v1/devices/10/samples/100` where there is a device with id 10 but no sample with id 100.

soren5 replied:
I think that specific route might not be documented in the API Reference, but I definitely agree that this isn't the most satisfying solution. 
This was the best I came up with after looking through the string methods for a bit, do you know what kind of tools I can use to improve this?

hmatalonga replied:
The endpoint actually doesn't exist was such an example :wink: 

I would keep it simple, don't even need to be so explicit about the ids, show instead a more generic message.

However we can take advantage of the `ModelNotFoundException` methods:

https://laravel.com/api/5.5/Illuminate/Database/Eloquent/ModelNotFoundException.html

hmatalonga commented on app/Exceptions/Handler.php:
```diff
@@ -49,18 +51,43 @@ public function report(Exception $exception)
+                        'message' => "The rate limit has been exceeded, please wait before sending more requests",
+                    ], 429);
+                } else {
+                    //Maybe this should be logged?
```
Yes, you are right any unhandled exceptions should be logged for later implementation. :+1:

soren5 replied:
I thought a bit more about this and I believe every exception is being logged at the moment, it should be changed so only unexpected errors get logged?

hmatalonga replied:
Actually we can even improve the logs by connecting to a 3rd party service like Sentry, which I think is a very reasonable idea for the near future.

Besides the info from the exception itself, in some cases we might need to add something more, we'll have to go case by case.

Logs should be handled on the `report` method.

https://laravel.com/docs/5.5/errors#report-method

At the moment, we don't have an easy way to go through the error logs since everything is still in a single text file.

soren5 replied:
I totally agree that logging should be improved soon, maybe another issue should be created to address that? I'm not sure how complex integration with a 3rd party is but it might keep things more organized

hmatalonga replied:
The service is already prepared to be integrated with Laravel apps, it is quite straightforward.

I have opened an issue for this matter #10

hmatalonga reviewed (changes requested):
Overall, this looks great! Nice job! :smiley:

I would still like to do some refactoring, clearing out some if statements by switching to a switch statement, no pun intended :stuck_out_tongue:

```php
$class = get_class($objectToTest);

switch($class) {  
    case TreeRequest::class:
        echo "tree request";
        break;
    case GroundRequest::class:
        echo "ground request";
        break;
}
```

Also you can, see on more recent versions of 5.5.*, they made some changes to the `Handler.php` class.

https://github.com/laravel/laravel/blob/5.5/app/Exceptions/Handler.php

The goal, is to override methods from `Illuminate\Foundation\Exceptions\Handler` when possible since there is already a place designed to add further logic to a given scenario.

Keeping the code as clean and readable as we can for future releases :wink: 

At the end, don't forget to run `composer format` before commiting the changes.

soren5 committed:
refactor:  clean up conditions

improve condition to catch api exceptions
restructured to avoid nested conditionals

soren5 committed:
refactor: switch case

substitute if else chain by a switch case

soren5 committed:
refactor: too many requests detection

change method used to distinguish too many requests exception from other
HttpExceptions

soren5 committed:
chore: format changes

format latest changes with composer format

soren5 committed:
feat: add error code 422

handle HttpResponseException with status code 422 and give
appropriate error message.

Issue #6

soren5 commented:
There is no API route that can throw a 422 error yet as far as I know so the last commit is untested.
That being said I found 2 ways to handle this exception:

1. By default laravel will throw a HttpResponseException with status code 422 when there is a problem validating as seen on : 
https://laravel.com/api/5.2/Illuminate/Foundation/Http/FormRequest.html#method_failedValidation
https://stackoverflow.com/questions/37146559/why-am-i-getting-a-422-error-code

2. I also found some links recommending to override the failedValidation method mentioned above to throw the more adequate ValidationException, here is an example: https://medium.com/@Sirolad/laravel-5-5-api-form-request-validation-errors-d49a85cd29f2

I decided to go with option 1 as it seems to be the most consistent with the way 429 was handled, option 2 also has the problem that it will force the failedValidation method to be overriden down the road.

hmatalonga commented on app/Exceptions/Handler.php:
```diff
@@ -43,24 +46,53 @@ public function report(Exception $exception)
+        $exceptionClass = get_class($exception);
+        switch ($exceptionClass) {
+            case NotFoundHttpException::class:
+                return response()->json(['message' => 'Requested route is invalid.',], 400);
```
I believe the proper status code for this case is *404 Not Found*.

After this minor fix, I think we are ready to finish this PR :wink:

soren5 replied:
Done! I'll be creating a PR for sentry integration soon

soren5 committed:
fix: change NotFoundHttpException response code

switch status code from 400 to 404 on NotFoundHttpException
tweak response messages

hmatalonga commented:
Closes #5, #6, #7
